debugger/src/debugger.rs

```diff
@@ -403,6 +408,76 @@ impl DebugSession {
     }
 }
 
+/// Manage the active RTT target for a specific DebugSession, as well as provide methods to reliably move RTT from target, through the debug_adapter, to the client.
+pub(crate) struct DebuggerRttTarget {
+    /// The  connection to RTT on the target
+    target_rtt: rtt::RttActiveTarget,
+    /// Some status fields and methods to ensure conitnuity in flow of data from target to debugger to client.
+    debugger_rtt_channels: Vec<DebuggerRttChannel>,
+}
+
+impl DebuggerRttTarget {
+    pub fn process_rtt_data<P: ProtocolAdapter>(
+        &mut self,
+        debug_adapter: &mut DebugAdapter<P>,
+        target_core: &mut Core,
+    ) -> bool {
+        let mut at_least_one_channel_had_data = false;
+        for debugger_rtt_channel in self.debugger_rtt_channels.iter_mut() {
+            if debugger_rtt_channel.send_rtt_data(target_core, debug_adapter, &mut self.target_rtt)
+            {
+                at_least_one_channel_had_data = true;
+            }
+        }
+        at_least_one_channel_had_data
+    }
+}
+
+pub(crate) struct DebuggerRttChannel {
+    pub(crate) channel_number: usize,
+    // We will not poll target RTT channels until we have confirmation from the client that the output window has been opened.
+    pub(crate) has_client_window: bool,
+}
+impl DebuggerRttChannel {
+    /// Poll and retrieve data from the target, and send it to the client, depending on the state of `hasClientWindow`.
+    /// Doing this selectively ensures that we don't pull data from target buffers until we have a output window, and also helps us drain buffers after the target has entered a `is_halted` state.
+    /// Errors will be reported back to the `debug_adapter`, and the return `bool` value indicates whether there was available data that was processed.
+    pub(crate) fn send_rtt_data<P: ProtocolAdapter>(
+        &mut self,
+        core: &mut Core,
+        debug_adapter: &mut DebugAdapter<P>,
+        rtt_target: &mut rtt::RttActiveTarget,
+    ) -> bool {
+        if self.has_client_window {
+            rtt_target
+                .active_channels
+                .iter_mut()
+                .find(|active_channel| {
+                    if let Some(channel_number) = active_channel.number() {
+                        channel_number == self.channel_number
+                    } else {
+                        false
+                    }
+                })
+                .and_then(|rtt_channel| {
+                    rtt_channel.get_rtt_data(core, rtt_target.defmt_state.as_ref())
```

Comment:
```suggestion
                    rtt_channel.get_rtt_data(core, &rtt_target.defmt_state)
```

I think this should work? :) Sometimes `as_ref()` is better (for chaining more methods) but here `&` would be more concise imo.

Reply:
Not this time. The `&` returns a reference to the `Option`, when I (and the compiler) want a reference to the contents of the `Option`.